src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceSetDeserializer.cs

```diff
@@ -93,20 +94,36 @@ public sealed override object ReadInline(object item, IEdmTypeReference edmType,
                 throw Error.Argument("edmType", SRResources.TypeMustBeResourceSet, edmType.ToTraceString());
             }
 
-            ODataResourceSetWrapper resourceSet = item as ODataResourceSetWrapper;
+            ODataResourceSetWrapperBase resourceSet = item as ODataResourceSetWrapperBase;
             if (resourceSet == null)
             {
-                throw Error.Argument("item", SRResources.ArgumentMustBeOfType, typeof(ODataResourceSetWrapper).Name);
+                throw Error.Argument("item", SRResources.ArgumentMustBeOfType, typeof(ODataResourceSetWrapperBase).Name);
             }
 
             // Recursion guard to avoid stack overflows
             RuntimeHelpers.EnsureSufficientExecutionStack();
 
             IEdmStructuredTypeReference elementType = edmType.AsCollection().ElementType().AsStructured();
 
+            IEdmEntityType actualType = edmType.AsCollection().ElementType().Definition as IEdmEntityType;
```

Comment:
this is only used if resourceSet.ResourceSetType == ResourceSetType.DeltaResouceSet, so move it into that block.